src/Microsoft.Health.Api/Features/HealthChecks/CachedHealthCheck.cs

```diff
@@ -10,91 +10,133 @@
+
+                HealthCheckResult result;
+                try
+                {
+                    IHealthCheck check = _healthCheck.Invoke(scope.ServiceProvider);
+                    result = await check.CheckHealthAsync(context, cancellationToken).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException oce) when (oce.CancellationToken == cancellationToken)
                 {
-                    try
-                    {
-                        IHealthCheck check = _healthCheck.Invoke(scope.ServiceProvider);
-                        _lastResult = await check.CheckHealthAsync(context, cancellationToken).ConfigureAwait(false);
-                        _lastChecked = Clock.UtcNow;
-                    }
-                    catch (Exception ex)
-                    {
-                        _logger.LogError(ex, "Health check failed to complete");
-                        _lastResult = HealthCheckResult.Unhealthy(Resources.FailedHealthCheckMessage);
-                        _lastChecked = Clock.UtcNow;
-                    }
+                    _logger.LogWarning(oce, $"Cancellation was requested for {nameof(CheckHealthAsync)}.");
+                    throw;
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogError(ex, "Health check failed to complete.");
+                    result = HealthCheckResult.Unhealthy(Resources.FailedHealthCheckMessage); // Do not pass error to caller
                 }
 
-                return _lastResult;
+                return UpdateCache(result).Value;
             }
             finally
             {
                 _semaphore.Release();
             }
         }
 
+        private CachedHealthCheckResult UpdateCache(HealthCheckResult result)
+        {
+            // Only update the cache if:
+            // (1) The current result has expired OR
+            // (2) The current result is stale and the cache has been configured to save failed results OR
+            // (3) The current result is stale and the new result is healthy
+            DateTimeOffset currentTime = Clock.UtcNow;
+            if (currentTime >= _cachedResult.ExpireTime || _options.CacheFailure || result.Status != HealthStatus.Unhealthy)
+            {
+                _cachedResult = new CachedHealthCheckResult
+                {
+                    ExpireTime = currentTime + _options.Expiry,
+                    StaleTime = currentTime + _options.Expiry - _options.RefreshOffset,
+                    Value = result,
+                };
+            }
+
+            return _cachedResult;
+        }
+
+        private bool IsCacheFresh(DateTimeOffset currentTime)
+            => currentTime < _cachedResult.StaleTime && (_options.CacheFailure || _cachedResult.Value.Status != HealthStatus.Unhealthy);
+
         public void Dispose()
         {
-            Dispose(true);
+            _semaphore.Dispose();
```

Comment:
Is there a chance for the `_semaphore` to be null? (Just asking because previously that was checked)

Reply:
Nope. It gets initialized before the ctor